draft-thomson-webpush-vapid.md

```diff
@@ -132,211 +134,279 @@ The words "MUST", "MUST NOT", "SHOULD", and "MAY" are used in this document.
+Proxy-Authorization header fields.
 
-The corresponding public key is included in a JSON Web Key (JWK) [RFC7517].
-This would be included in either a newly-defined "jwk" parameter of the
-Crypto-Key header field [I-D.ietf-httpbis-encryption-encoding]; alternatively, the
-"p256ecdsa" parameter defined in [I-D.thomson-http-content-signature] could be
-used to transport a raw key.
+This authentication scheme does not require a challenge.  Clients are able to
+generate the Authorization header field without any additional information from
+a server.  Therefore, a challenge for this authentication scheme MUST NOT be
+sent in a WWW-Authenticate header field.
 
-For voluntarily-provided contact details, a separate header field could be used
-(as in {{from}}) or the JWT could include claims about identity.  For the
-latter, the "sub" (subject) claim could include a contact URI for the
-application server.
+A single parameter is defined for this challenge.  A `keyid` parameter includes
+an identifier for the key that was used to sign the JWT.  This parameter MAY be
+omitted if the recipient can be expected to locate the correct key without this
+assistance.  For instance, if there is exactly one Crypto-Key header field in
+the same request that contains a key of the right type, the `keyid` can be
+safely omitted.
+
+All unknown or unsupported parameters to "WebPush" authentication credentials
+MUST be ignored.  The `realm` parameter is ignored for this authentication
+scheme.
 
-The JWT MUST use a JSON Web Signature (JWS) [RFC7515].  Both the JWS and JWK
-MUST use an elliptic curve digital signature (ECDSA) key on the NIST P-256 curve
-[FIPS186].
 
-A JWT also offers the option of including contact information as an additional
-claim.  An "iss" (Issuer) claim can include a contact URI: either a "mailto:"
-(email) [RFC6068] or an "https:" [RFC2818] SHOULD be used.
+# Public Key Representation {#key}
 
+In order for the push service to be able to validate the JWT, it needs to learn
+the public key of the application server.  A `p256ecdsa` parameter is defined
+for the Crypto-Key header field [I-D.ietf-httpbis-encryption-encoding] to carry
+this information.
 
-## Contact Information Header Field {#from}
+The `p256ecdsa` parameter includes an elliptic curve digital signature algorithm
+(ECDSA) public key [FIPS186] in uncompressed form [X9.62] that is encoded using
+the URL- and filename-safe variant of base-64 [RFC4648] with padding removed.
 
-Contact information for an application server could be included in a header
-field, either directly (e.g., a From header field, Section 5.5.1 of [RFC7231]),
-or by reference (e.g., a new "contact" link relation [RFC5988] that identified a
-vCard [RFC6350]).  Note that a From header field is limited to email addresses.
+Note that with push message encryption [I-D.ietf-webpush-encryption], this
+results in two values in the Crypto-Key header field, one with the a `p256dh`
+key and another with a `p256ecdsa` key.
```

Comment:
So the push service would consume "p256ecdsa", and pass through "p256dh" to the client. Today the push service just passes through the value to the client, so it'll have to do parsing. (Which should be fine.)